jbmc/unit/java_bytecode/java_object_factory/struct_tag_types.cpp

```diff
@@ -0,0 +1,204 @@
+    pointer_selector);
+
+  return created_code;
+}
+
+SCENARIO(
+  "java_object_factory_tag_types",
+  "[core][java_bytecode][java_object_factory]")
+{
+  GIVEN("Some classes with fields")
+  {
+    register_language(new_java_bytecode_language);
+    symbol_tablet symbol_table =
+      load_java_class("Root", "./java_bytecode/java_object_factory");
+
+    WHEN("Creating a nondet 'A' object")
+    {
+      struct_tag_typet A_type("java::A");
+      struct_tag_typet B_type("java::B");
+      struct_tag_typet C_type("java::C");
+
+      const auto a_pointer = java_reference_type(A_type);
+      code_blockt created_code =
+        initialise_nondet_object_of_type(a_pointer, symbol_table);
+
+      THEN("An A, a B and a C object should be allocated")
+      {
+        REQUIRE(contains_decl_of_type(created_code, A_type));
+        REQUIRE(contains_decl_of_type(created_code, B_type));
+        REQUIRE(contains_decl_of_type(created_code, C_type));
+      }
+
+      THEN("No raw struct expressions should appear")
+      {
+        REQUIRE(!uses_raw_struct_types(created_code));
+      }
+    }
+
+    WHEN("Creating a nondet 'HasArray' object")
+    {
+      struct_tag_typet HasArray_type("java::HasArray");
+      struct_tag_typet D_type("java::D");
+
+      const auto HasArray_pointer = java_reference_type(HasArray_type);
+      code_blockt created_code =
+        initialise_nondet_object_of_type(HasArray_pointer, symbol_table);
+
+      THEN("A HasArray object should be created")
+      {
+        REQUIRE(contains_decl_of_type(created_code, HasArray_type));
+      }
+
+      THEN("No raw struct expressions should appear")
+      {
+        REQUIRE(!uses_raw_struct_types(created_code));
+      }
+    }
+
+    WHEN("Creating a nondet 'HasEnum' object")
+    {
+      struct_tag_typet HasEnum_type("java::HasEnum");
+      struct_tag_typet E_type("java::E");
+
+      const auto HasEnum_pointer = java_reference_type(HasEnum_type);
+      code_blockt created_code =
+        initialise_nondet_object_of_type(HasEnum_pointer, symbol_table);
+
+      THEN("A HasEnum object should be allocated")
+      {
+        REQUIRE(contains_decl_of_type(created_code, HasEnum_type));
+      }
+
+      THEN("No raw struct expressions should appear")
+      {
+        REQUIRE(!uses_raw_struct_types(created_code));
+      }
+    }
+
+    WHEN("Creating a nondet 'HasGeneric' object")
+    {
+      struct_tag_typet HasGeneric_type("java::HasGeneric");
+      irep_idt generic_id = "java::Generic";
+      irep_idt other_generic_id = "java::OtherGeneric";
+
+      const auto HasGeneric_pointer = java_reference_type(HasGeneric_type);
+      code_blockt created_code =
+        initialise_nondet_object_of_type(HasGeneric_pointer, symbol_table);
+
+      THEN("An HasGeneric, Generic and OtherGeneric object should be allocated")
+      {
+        REQUIRE(contains_decl_of_type(created_code, HasGeneric_type));
+        REQUIRE(contains_decl_with_id(created_code, generic_id));
```

Comment:
:pick: I think this name is misleading: I was expecting it to be the identifier of the declaration, rather than the tag of the type, perhaps a better name would be `contains_decl_with_struct_tag`?

Reply:
Also, why do you handle these via just checking tags rather than through the actual type?

Reply:
Laziness -- they're generics so they have decorators (so they're not `==` to a plain `struct_tag_typet`). Explained in a comment; I'll leave it to the generics tests to verify the structure of generic types.